### PR TITLE
Create a state-change transform type for the t128_transform processor

### PR DESCRIPTION
**Pull request by gregschrock**
## Description

The state change transform will detect changes in a field and only pass that field along if a change took place. It will pass the first observed value (start-up) and the first value observed after "expiration" if the expiration is specified.

### CODE REVIEW

File: plugins/processors/t128_transform/t128_transform.go [hunk 1]
```diff
@@ -77,36 +81,35 @@ func (r *T128Transform) Apply(in ...telegraf.Metric) []telegraf.Metric {
...
+					value: nil,
 				}
 			}
 
-			if (target.matchesSource && !itemAdded) || (!target.matchesSource && r.RemoveOriginal) {
+			expired := !point.Time().Before(observed.expires)
+
+			itemTransformed := false
+			value, err := r.transform(expired, observed.timestamp, point.Time(), observed.value, field.Value)
+			if err != nil {
```
Comment (agrawalkaushik) [inline plugins/processors/t128_transform/t128_transform.go:123]: I know this is meant for good practice error handling but any practical reasons why the `transform` function would fail?

File: plugins/processors/t128_transform/t128_transform_state_change_test.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+		Name            string
+		Fields          map[string]string
+		LastSample      map[string]interface{}
+		CurrentSample   map[string]interface{}
+		TimeDelta       time.Duration
+		RemoveOriginal  bool
+		RemainingFields []string
+	}{
+		{
+			Name:            "remove-matching-new",
```
Comment (agrawalkaushik) [inline plugins/processors/t128_transform/t128_transform_state_change_test.go]: is `remove-` referring to the `RemoveOriginal` part coz its slightly confusing when looking at `RemainingFields`

File: plugins/processors/t128_transform/t128_transform_state_change_test.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+			LastSample:      map[string]interface{}{},
+			CurrentSample:   map[string]interface{}{"/state": 50},
+			TimeDelta:       5 * time.Second,
+			RemoveOriginal:  true,
+			RemainingFields: []string{"/state"},
+		},
+		{
+			Name:            "remove-matching-existing-same",
+			Fields:          map[string]string{"/state": "/state"},
+			LastSample:      map[string]interface{}{"/state": 45},
```
Comment (agrawalkaushik) [inline plugins/processors/t128_transform/t128_transform_state_change_test.go]: can we throw in some non-numeric values in some of the test cases for coverage

File: plugins/processors/t128_transform/t128_transform.go [hunk 4]
```diff
@@ -12,13 +12,17 @@ import (
 )
```
Comment (agrawalkaushik) [inline plugins/processors/t128_transform/t128_transform.go:12]: Are we including the "old" and "new" values in the data somewhere?

**Review by agrawalkaushik — APPROVED** [review]

File: plugins/processors/t128_transform/t128_transform.go (hunk 1 above)
Comment (gregschrock) [reply plugins/processors/t128_transform/t128_transform.go:123]: It can fail on rate and diff if the field is not numeric

File: plugins/processors/t128_transform/t128_transform_state_change_test.go (hunk 2 above)
Comment (gregschrock) [reply plugins/processors/t128_transform/t128_transform_state_change_test.go]: Yes, this was an attempt to identify all the combinations of the configuration. It's a replication of what's done for diff and rate.

File: plugins/processors/t128_transform/t128_transform_state_change_test.go (hunk 3 above)
Comment (gregschrock) [reply plugins/processors/t128_transform/t128_transform_state_change_test.go]: `TestStateChangeSendFirstSample` and `TestStateChangeSendsOnExpired` were intended to cover the non-numeric case.

File: plugins/processors/t128_transform/t128_transform.go (hunk 4 above)
Comment (gregschrock) [reply plugins/processors/t128_transform/t128_transform.go:12]: This does not report the old value explicitly. I didn't think it makes sense in the context of the existing transforms and thought we'd concluded that we're just interested in suppressing repetitive output rather than identifying change edges per se.

Comment (agrawalkaushik) [reply plugins/processors/t128_transform/t128_transform.go:12]: I had a different recollection - we had settled on including the old value (at least for the state case) as a tag so the user can decide whether they need it or not. For the ARP state change even for example, we should report the transition - the old state and the new state - that information is very much relevant

File: plugins/processors/t128_transform/t128_transform.go [hunk 5]
```diff
@@ -77,36 +81,35 @@ func (r *T128Transform) Apply(in ...telegraf.Metric) []telegraf.Metric {
...
+			itemTransformed := false
+			value, err := r.transform(expired, observed.timestamp, point.Time(), observed.value, field.Value)
+			if err != nil {
+				r.Log.Warnf("excluding failed transform: %v", err)
+			} else if value != nil {
+				itemTransformed = true
+				point.AddField(target.key, value)
+			}
+
+			if (target.matchesSource && !itemTransformed) || (!target.matchesSource && r.RemoveOriginal) {
```
Comment (BenMatase) [inline plugins/processors/t128_transform/t128_transform.go:130]: There is a lot going on in this func so it a little hard to follow. Maybe some intermediate vars or helper funcs to help clarify the different parts? No biggie

**Review by BenMatase — APPROVED** [review]

File: plugins/processors/t128_transform/t128_transform.go (hunk 4 above)
Comment (gregschrock) [reply plugins/processors/t128_transform/t128_transform.go:12]: Okay, will revisit

- [x] report previous values

File: plugins/processors/t128_transform/t128_transform.go (hunk 5 above)
Comment (gregschrock) [reply plugins/processors/t128_transform/t128_transform.go:130]: I honestly tried to clean it up and there are enough moving pieces that it becomes more complicated. I'm going to leave it the way it is for now.

Comment (BenMatase) [reply plugins/processors/t128_transform/t128_transform.go:130]: That is fine. Thanks for trying

File: plugins/processors/t128_transform/t128_transform_state_change_test.go [hunk 6]
```diff
@@ -74,182 +74,254 @@ func TestStateChangeLeavesUnmarkedFieldsInTact(t *testing.T) {
...
+			Samples: []sample{
+				{"/non-state": 45},
+				{"/non-state": 50},
+			},
+			TimeDelta:      6 * time.Second,
+			RemoveOriginal: false,
+			Result:         sample{"/non-state": int64(50), "/state": int64(50)},
+		},
+
+		{
```
Comment (gregschrock) [inline plugins/processors/t128_transform/t128_transform_state_change_test.go:265]: This is where new tests start. The earlier changes are just converting to `Samples` and `Result` to make the test more flexible.

File: plugins/processors/t128_transform/t128_transform_test.go [hunk 7]
```diff
@@ -165,147 +165,228 @@ func TestLeavesUnmarkedFieldsInTact(t *testing.T) {
...
+			Samples: []sample{
+				{"/non-rate": 45},
+				{"/non-rate": 50},
+			},
+			TimeDelta:      6 * time.Second,
+			RemoveOriginal: false,
+			Result:         sample{"/non-rate": int64(50)},
 		},
+
 		{
```
Comment (gregschrock) [inline plugins/processors/t128_transform/t128_transform_test.go:313]: Again, this is where the new tests start. The earlier changes are just converting to a more flexible test model. That conversion is needed because some transforms won't have a "previous" value until the third data point and the old model only allowed for two samples. The change also checks the values of the fields rather than only their names.

File: plugins/processors/t128_transform/t128_transform_state_change_test.go [hunk 8]
```diff
@@ -74,182 +74,254 @@ func TestStateChangeLeavesUnmarkedFieldsInTact(t *testing.T) {
...
+			Name:           "previous-with-previous-value",
+			Fields:         map[string]string{"/state": "/state"},
+			PreviousFields: map[string]string{"/previous": "/state"},
+			Samples: []sample{
+				{"/state": "s1"},
+				{"/state": "s2"},
+			},
+			TimeDelta:      5 * time.Second,
+			RemoveOriginal: true,
+			Result:         sample{"/state": "s2", "/previous": "s1"},
```
Comment (agrawalkaushik) [inline plugins/processors/t128_transform/t128_transform_state_change_test.go:299]: Does this rely completely on the current and previous sample for the "previous" case? In other words if `current == previous` we would not produce anything but do we record that? I would imagine only the observed "previous" to be reported - but just want to confirm that behavior.

**Review by agrawalkaushik — APPROVED** [review]

File: plugins/processors/t128_transform/t128_transform_state_change_test.go (hunk 8 above)
Comment (gregschrock) [reply plugins/processors/t128_transform/t128_transform_state_change_test.go:299]: That's correct. But your question helped me identify a bug in a codepath I hadn't tested. When the transform should produce nothing, it returns a `nil` value. That's currently being recorded as the previous which isn't correct in some cases.
- [x] test the new corner case

**Review by salegria8 — APPROVED** [review]

**Review by jamesesilvia — APPROVED** [review]

**Review by BenMatase — APPROVED** [review]
